src/engine/Operation.cpp

```diff
@@ -19,3 +39,94 @@ vector<string> Operation::collectWarnings() const {
 
   return res;
 }
+
+// Get the result for the subtree rooted at this element.
+// Use existing results if they are already available, otherwise
+// trigger computation.
+shared_ptr<const ResultTable> Operation::getResult(bool isRoot) {
+  ad_utility::Timer timer;
+  timer.start();
+  auto& cache = _executionContext->getQueryTreeCache();
+  const string cacheKey = asString();
```

Comment:
This function was mostly only moved from the .h to the .cpp file because of cyclic dependencies.
New Code starts at this marker.